### DISCUSSION TITLE
library_adjust option in symtabAPI

### DISCUSSION THREAD

**Discussion by sagsango** [dyninst/dyninst#1764] [General]
In emitElf.C: template<class ElfTypes> emitElf<ElfTypes>::emitElf()
   ```
 //If we want to try a mode where we add the program headers to a library
    // that can be loaded anywhere, and put the program headers in the first
    // page (avoiding the kernel bug), then set library_adjust to getpagesize().
    // This will shift all addresses in the library down by a page, accounting
    // for the extra page for program headers.  This causes some significant
    // changes to the binary, and isn't well tested.

    library_adjust = 0;
    if (cannotRelocatePhdrs() && !movePHdrsFirst) {
        movePHdrsFirst = true;
        library_adjust = getpagesize();
    }
```
    
There is a library_adjust option that breaks the alignment of the .data section where the alignment is greater than the page size.

Any downside if I disable this? Any plan on fixing it?

**Comment by bigtrak** [comment]
>     library_adjust = 0;
>     if (cannotRelocatePhdrs() && !movePHdrsFirst) {
>         movePHdrsFirst = true;
>         library_adjust = getpagesize();
>     }
> ```
>     
> there is library_adjust option which breaks the alignmnet of .data section where alignmnet is greter that page size.
> 
> Any downside if I disable this? Any plan on fixing it? 

Can you identify the different alignment 
required for the platform you are using, and
where that information or change originates?

Bolo

**Comment by sagsango** [comment]
I'm running the glibc tests after instrumenting the glibc library. However, the test that checks for alignment is failing.

[Link](https://github.com/bminor/glibc/blob/master/elf/tst-align3.c) https://github.com/bminor/glibc/blob/master/elf/tst-align3.c

Where they are testing for the  alignment ALIGN = 0x200000

**Reply by sagsango** [reply]
Adding the logs:

```
# ./runner.sh elf/tst-align3
bar: 0x555555e00000
foo: 0x7ffff7a01000
error: tst-alignmod3.c:29: not true: is_aligned (&foo, ALIGN) == 0
error: 1 test failures
```

Original elf .data section:
 `[25] .data             PROGBITS        0000000000800000 600000 200004 00  WA  0   0 2097152`

Instrumented elf .data section
`[25] .data             PROGBITS        0000000000801000 601000 200004 00  WA  0   0 2097152`

**Comment by kupsch** [comment]
I can look at fixing this.  I can build glibc and the tst-align3 binary with its larger than the page size alignment.  I would be helpful if you could send a small test mutator and instruction on how to run it on the tst-align3 that creates the bad elf file so I can recreate the problem and verify the fix.

Thanks.

**Comment by sagsango** [comment]
This code can be used to test alignments, which create the "foo" variable in a foo section and the bar in the data section. After the instrumentation, alignment should be preserved.

```
#include <stdio.h>
#include <stdint.h>

#define ALIGN_foo (0x100000)
#define ALIGN_bar (0x200000)

#define TEST_VERIFY(expr) if (!(expr)) { \
    fprintf(stderr, "Test failed: %s\n", #expr); \
    return 1; \
}

const int __attribute__ ((section(".foo"), aligned(ALIGN_foo))) foo = 0x12345678;
int bar __attribute__ ((aligned (ALIGN_bar))) = 1;

// Helper function to check alignment
int is_aligned(void *ptr, size_t align) {
    return ((unsigned long long)ptr % align);
}



static int
do_test (void)
{
    printf("bar: %p\n", (void*)&bar);
    printf("foo: %p\n", (void*)foo);
    TEST_VERIFY(is_aligned(&bar, ALIGN_bar) == 0);
    TEST_VERIFY(is_aligned(&foo, ALIGN_foo) == 0);

    return 0;
}

int main(void) {
    return do_test();
}
```


$ readelf -aW test
```
Section Headers:
  [Nr] Name              Type            Address          Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            0000000000000000 000000 000000 00      0   0  0
  [ 1] .interp           PROGBITS        0000000000000388 000388 00001c 00   A  0   0  1
  [ 2] .note.gnu.property NOTE            00000000000003a8 0003a8 000020 00   A  0   0  8
  [ 3] .note.gnu.build-id NOTE            00000000000003c8 0003c8 000024 00   A  0   0  4
  [ 4] .note.ABI-tag     NOTE            00000000000003ec 0003ec 000020 00   A  0   0  4
  [ 5] .gnu.hash         GNU_HASH        0000000000000410 000410 000028 00   A  6   0  8
  [ 6] .dynsym           DYNSYM          0000000000000438 000438 0000d8 18   A  7   1  8
  [ 7] .dynstr           STRTAB          0000000000000510 000510 00008c 00   A  0   0  1
  [ 8] .gnu.version      VERSYM          000000000000059c 00059c 000012 02   A  6   0  2
  [ 9] .gnu.version_r    VERNEED         00000000000005b0 0005b0 000020 00   A  7   1  8
  [10] .rela.dyn         RELA            00000000000005d0 0005d0 0000d8 18   A  6   0  8
  [11] .rela.plt         RELA            00000000000006a8 0006a8 000030 18  AI  6  25  8
  [12] .init             PROGBITS        0000000000001000 001000 00001b 00  AX  0   0  4
  [13] .plt              PROGBITS        0000000000001020 001020 000030 10  AX  0   0 16
  [14] .plt.got          PROGBITS        0000000000001050 001050 000010 10  AX  0   0 16
  [15] .plt.sec          PROGBITS        0000000000001060 001060 000020 10  AX  0   0 16
  [16] .text             PROGBITS        0000000000001080 001080 000255 00  AX  0   0 16
  [17] .fini             PROGBITS        00000000000012d8 0012d8 00000d 00  AX  0   0  4
  [18] .rodata           PROGBITS        0000000000002000 002000 000079 00   A  0   0  8
  [19] .foo              PROGBITS        0000000000100000 100000 000004 00   A  0   0 1048576
  [20] .eh_frame_hdr     PROGBITS        0000000000100004 100004 000054 00   A  0   0  4
  [21] .eh_frame         PROGBITS        0000000000100058 100058 000148 00   A  0   0  8
  [22] .init_array       INIT_ARRAY      0000000000101db0 100db0 000008 08  WA  0   0  8
  [23] .fini_array       FINI_ARRAY      0000000000101db8 100db8 000008 08  WA  0   0  8
  [24] .dynamic          DYNAMIC         0000000000101dc0 100dc0 0001f0 10  WA  7   0  8
  [25] .got              PROGBITS        0000000000101fb0 100fb0 000050 08  WA  0   0  8
  [26] .data             PROGBITS        0000000000200000 200000 200004 00  WA  0   0 2097152
  [27] .bss              NOBITS          0000000000400020 400004 000010 00  WA  0   0 32
  [28] .comment          PROGBITS        0000000000000000 400004 00002b 01  MS  0   0  1
  [29] .symtab           SYMTAB          0000000000000000 400030 0006c0 18     30  48  8
  [30] .strtab           STRTAB          0000000000000000 4006f0 000235 00      0   0  1
  [31] .shstrtab         STRTAB          0000000000000000 400925 00011f 00      0   0  1
  
```

**Reply by sagsango** [reply]
If we want to run the test program, we need newer versions of glibc, such as glibc-2.39, which is present in Ubuntu 24.04. (However, we don't need to run after the instrumentation because elf itself gives us the alignment details)


More: https://sourceware.org/bugzilla/show_bug.cgi?id=28676

**Reply by kupsch** [reply]
Sorry, I guess I wasn't clear.  I was able to create a program with larger than page size alignment that is causing the problem when rewritten by dyninst.  What would be useful would be a test program that uses dyninst to rewrite the binary resulting in incorrectly aligned segments.

**Reply by sagsango** [reply]
It does not matter what type of instrumentation is used; this problem arises whenever dynint generates a new binary.

For example, in https://github.com/dyninst/examples/blob/master/instrumentAFunction/instrumenting_a_function.cpp, I changed the mode in line 146 to open and used the above example code as the target.

Attaching the source code used to reproduce the issue: 
- instrumentation code:  [dyninst-instrumentation.txt](https://github.com/user-attachments/files/16072326/dyninst-instrumentation.txt)
- target binary code:  [InterestingProgram.txt](https://github.com/user-attachments/files/16072329/InterestingProgram.txt)

**Reply by sagsango** [reply]
@kupsch, any update on it?

**Reply by hainest** [reply]
@bbiiggppiigg @kupsch This sounds like it may be fixed by #2119.
